File: frontend/lib/data-sources/http/papers.ts
import type { EditorialDraftsResponse, PaperInsightsResponse, PapersListResponse } from "../../api-contracts.ts";
import type { PaperSearchParams } from "../../types.ts";
import { createHttpClient, type HttpDataSourceOptions, type HttpQueryParams } from "./shared.ts";
import type { PapersDataSource } from "../demo/papers.ts";

export function createHttpPapersDataSource(options: HttpDataSourceOptions): PapersDataSource {
  const client = createHttpClient(options);

  return {
    async listPapers() {
      const response = await client.get<PapersListResponse>("papers");

      return response.items.map((item) => item.paper);
    },
    async listInsights() {
      const response = await client.get<PaperInsightsResponse>("papers/insights");

      return [...response.items];
    },
    async listEditorialDrafts() {
      const response = await client.get<EditorialDraftsResponse>("papers/editorial-drafts");

      return [...response.items];
    },
    async searchPapers(params: PaperSearchParams) {
      const query: HttpQueryParams = {};

      if (params.q) {
        query.q = params.q;
      }

      if (params.source && params.source !== "all") {
        query.source = params.source;
      }

      if (params.category) {
        query.category = params.category;
      }

      if (params.venue) {
        query.venue = params.venue;
      }

      if (params.hasInsight !== undefined) {
        query.hasInsight = params.hasInsight;
      }

      if (params.hasDraft !== undefined) {
        query.hasDraft = params.hasDraft;
      }

      if (params.page && params.pageSize) {
        query.limit = params.pageSize;
        query.offset = (params.page - 1) * params.pageSize;
      }

      const response = await client.get<PapersListResponse>("papers", query);

      return {
        items: response.items.map((item) => item.paper),
        total: response.total,
      };
    },
  };
}

File: app/api/routes/papers.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query, Request

from app.api.schemas import EditorialDraftsResponse, PaperDetailResponse, PaperInsightsResponse, PapersListResponse, create_envelope
from app.api.services.editorial_workflow import list_drafts
from app.api.services.read_models import get_paper_detail, list_paper_insights, list_papers

router = APIRouter(prefix="/papers", tags=["papers"])


@router.get("")
def get_papers(
    request: Request,
    q: str = "",
    source: str | None = None,
    category: str | None = None,
    venue: str | None = None,
    hasInsight: bool | None = None,
    hasDraft: bool | None = None,
    limit: int | None = Query(default=None, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
) -> dict:
    db = request.app.state.db
    editorial_root = request.app.state.editorial_root
    items, total = list_papers(
        db,
        editorial_root,
        q=q,
        source=source,
        category=category,
        venue=venue,
        has_insight=hasInsight,
        has_draft=hasDraft,
        limit=limit,
        offset=offset,
    )
    return create_envelope(PapersListResponse(items=items, total=total, appliedQuery=q)).model_dump(by_alias=True)


@router.get("/insights")
def get_paper_insights(request: Request) -> dict:
    db = request.app.state.db
    items = list_paper_insights(db)
    return create_envelope(PaperInsightsResponse(items=items, total=len(items))).model_dump(by_alias=True)


@router.get("/editorial-drafts")
def get_editorial_drafts(request: Request) -> dict:
    items = list_drafts(request.app.state.db)
    return create_envelope(EditorialDraftsResponse(items=items, total=len(items))).model_dump(by_alias=True)


@router.get("/{paper_id}")
def get_paper(request: Request, paper_id: int) -> dict:
    db = request.app.state.db
    payload = get_paper_detail(db, paper_id)
    if payload is None:
        raise HTTPException(status_code=404, detail="paper not found")
    return create_envelope(PaperDetailResponse(**payload.model_dump(by_alias=True))).model_dump(by_alias=True)
